proc/cut.go

```diff
@@ -5,74 +5,268 @@ import (
+// to be adjacent.  Alternatively we could re-order provided fields
+// so the output record can be constructed efficiently, though we don't
+// do this now since it might confuse users who expect to see output
+// fields in the order they specified.
+func CompileCutProc(c *Context, parent Proc, node *ast.CutProc) (*Cut, error) {
+	seenRecords := make(map[string]bool)
+	fieldInfos := make([]fieldInfo, 0, len(node.Fields))
+	var currentRecord []string
+	for i, field := range node.Fields {
+		resolver, err := expr.CompileFieldExpr(field)
+		if err != nil {
+			return nil, err
+		}
+
+		names, err := split(field)
+		if err != nil {
+			return nil, err
+		}
+
+		// Grab everything except the leaf field name and see if
+		// it has changed from the previous field.  If it hasn't,
+		// things are simple but if it has, we need to carefully
+		// figure out which records we are stepping in and out of.
+		record := names[:len(names)-1]
+		var containerBegins []string
+		if !sameRecord(record, currentRecord) {
+			// currentRecord is what nested record the zval.Builder
+			// is currently working on, record is the nested
+			// record for the current field.  First figure out
+			// what (if any) common parents are shared.
+			l := len(currentRecord)
+			if len(record) < l {
+				l = len(record)
+			}
+			pos := 0
+			for pos < l {
+				if record[pos] != currentRecord[pos] {
+					break
+				}
+				pos += 1
+			}
+
+			// Note any previously encoded records that are
+			// now finished.
+			if i > 0 {
+				fieldInfos[i-1].containerEnds = len(currentRecord) - pos
+			}
+
+			// Validate any new records that we're starting
+			// (i.e., ensure that we didn't handle fields from
+			// the same record previously), then record the names
+			// of all these records.
+			for pos2 := pos; pos2 < len(record); pos2++ {
+				recname := strings.Join(record[:pos2+1], ".")
+				_, seen := seenRecords[recname]
+				if seen {
+					return nil, fmt.Errorf("All cut fields in record %s must be adjacent", recname)
+				}
+				seenRecords[recname] = true
+				containerBegins = append(containerBegins, record[pos2])
+			}
+			currentRecord = record
+		}
+		fullname := strings.Join(names, ".")
+		fname := names[len(names)-1]
+		fieldInfos = append(fieldInfos, fieldInfo{resolver, fname, fullname, containerBegins, 0})
+	}
+	fieldInfos[len(fieldInfos)-1].containerEnds = len(currentRecord)
+
 	return &Cut{
-		Base:   Base{Context: c, Parent: parent},
-		fields: fields,
-		cutmap: make(map[int]*zson.Descriptor),
+		Base:    Base{Context: c, Parent: parent},
+		fields:  fieldInfos,
+		cutmap:  make(map[int]*zson.Descriptor),
+		builder: zval.NewBuilder(),
+	}, nil
+}
+
+// Split an ast.FieldExpr representing a chain of record field references
+// into a list of strings representing the names.
+// E.g., "x.y.z" -> ["x", "y", "z"]
+func split(node ast.FieldExpr) ([]string, error) {
+	switch n := node.(type) {
+	case *ast.FieldRead:
+		return []string{n.Field}, nil
+	case *ast.FieldCall:
```

Comment:
this is made impossible by the current grammar right? if so i would suggest relaxing the grammar, as it's more helpful to get the error message below than "parse error: 1:1 (0): no match found".

Reply:
(where by "current" i really meant "new as of this PR"....)

Reply:
That's how I wrote it at first, but the javascript parser gives a better error so in this way, the desktop app catches an invalid proc (e.g., `cut foo[2]`) earlier and with a reasonable error.  I don't feel strongly about this though, happy to change it if you prefer.

Reply:
good points... i feel less strongly now i understand this context, so feel free not to change it...